Engine/Fuego/Windows/WindowWin.cpp

```diff
@@ -1,26 +1,54 @@
-#include "WindowWin.h"
+﻿#include "WindowWin.h"
 
 namespace Fuego
 {
-LRESULT CALLBACK WindowProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+std::unordered_map<HWND, WindowWin*> WindowWin::hwndMap;
+
+LRESULT CALLBACK WindowWin::WindowProcStatic(HWND hWnd, UINT uMsg,
+                                             WPARAM wParam, LPARAM lParam)
+{
+    auto _this = WindowWin::hwndMap.find(hWnd);
+
+    if (_this != WindowWin::hwndMap.end() && _this->second)
+    {
+        return _this->second->WindowProc(hWnd, uMsg, wParam, lParam);
+    }
+    else
+    {
+        return DefWindowProc(hWnd, uMsg, wParam, lParam);
+    }
+}
+
+LRESULT WindowWin::WindowProc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam)
 {
-    switch (uMsg)
+    switch (msg)
     {
+        case WM_KEYDOWN:
+        {
+            int repeatCount = (lparam >> 16) & 0xFF;
+            m_EventQueue.PushEvent(std::make_unique<KeyPressedEvent>(
+                static_cast<int>(wparam), repeatCount));
+            break;
+        }
+        case WM_KEYUP:
+            m_EventQueue.PushEvent(
+                std::make_unique<KeyReleasedEvent>(static_cast<int>(wparam)));
+            break;
+
         case WM_CLOSE:
-            DestroyWindow(hWnd);
+            m_EventQueue.PushEvent(std::make_unique<WindowCloseEvent>());
             break;
-        case WM_DESTROY:
```

Comment:
We should think how to respond to this event

Reply:
We will, not in this commit